// backend/src/services/auth.ts
import jwt from 'jsonwebtoken';
import { PrismaClient, User } from '@prisma/client';

const prisma = new PrismaClient();

export interface TokenPayload {
  userId: string;
  email: string;
  role: string;
  familyId: string;
}

export interface TokenPair {
  accessToken: string;
  refreshToken: string;
  expiresIn: number;
}

export class AuthService {
  private static readonly ACCESS_TOKEN_EXPIRES_IN = '15m';
  private static readonly REFRESH_TOKEN_EXPIRES_IN = '7d';

  /**
   * Generate JWT tokens for user
   */
  static generateTokens(user: User): TokenPair {
    const payload: TokenPayload = {
      userId: user.id,
      email: user.email,
      role: user.role,
      familyId: user.familyId,
    };

    const accessToken = jwt.sign(
      payload,
      process.env.JWT_SECRET!,
      { expiresIn: this.ACCESS_TOKEN_EXPIRES_IN }
    );

    const refreshToken = jwt.sign(
      { userId: user.id, type: 'refresh' },
      process.env.JWT_REFRESH_SECRET!,
      { expiresIn: this.REFRESH_TOKEN_EXPIRES_IN }
    );

    return {
      accessToken,
      refreshToken,
      expiresIn: 15 * 60, // 15 minutes in seconds
    };
  }

  /**
   * Verify and decode access token
   */
  static verifyAccessToken(token: string): TokenPayload | null {
    try {
      const decoded = jwt.verify(token, process.env.JWT_SECRET!) as TokenPayload;
      return decoded;
    } catch (error) {
      console.error('Access token verification failed:', error);
      return null;
    }
  }

  /**
   * Verify and decode refresh token
   */
  static verifyRefreshToken(token: string): { userId: string; type: string } | null {
    try {
      const decoded = jwt.verify(token, process.env.JWT_REFRESH_SECRET!) as any;
      if (decoded.type !== 'refresh') {
        throw new Error('Invalid token type');
      }
      return decoded;
    } catch (error) {
      console.error('Refresh token verification failed:', error);
      return null;
    }
  }

  /**
   * Refresh access token using refresh token
   */
  static async refreshAccessToken(refreshToken: string): Promise<TokenPair | null> {
    const decoded = this.verifyRefreshToken(refreshToken);
    if (!decoded) {
      return null;
    }

    // Get user from database
    const user = await prisma.user.findUnique({
      where: { id: decoded.userId },
    });

    if (!user) {
      return null;
    }

    // Generate new tokens
    return this.generateTokens(user);
  }

  /**
   * Get user from access token
   */
  static async getUserFromToken(token: string): Promise<User | null> {
    const payload = this.verifyAccessToken(token);
    if (!payload) {
      return null;
    }

    const user = await prisma.user.findUnique({
      where: { id: payload.userId },
      include: { family: true },
    });

    return user;
  }

  /**
   * Check if user has required role
   */
  static hasRole(user: User, requiredRoles: string[]): boolean {
    return requiredRoles.includes(user.role);
  }

  /**
   * Check if user can access family data
   */
  static canAccessFamily(user: User, familyId: string): boolean {
    return user.familyId === familyId;
  }

  /**
   * Check if user can access expense (own or family member)
   */
  static async canAccessExpense(user: User, expenseUserId: string): Promise<boolean> {
    // User can always access their own expenses
    if (user.id === expenseUserId) {
      return true;
    }

    // Parents can access all family expenses
    if (user.role === 'FATHER' || user.role === 'MOTHER') {
      const expenseUser = await prisma.user.findUnique({
        where: { id: expenseUserId },
      });
      return expenseUser?.familyId === user.familyId;
    }

    return false;
  }

  /**
   * Check if user can manage budgets (Father role only)
   */
  static canManageBudgets(user: User): boolean {
    return user.role === 'FATHER';
  }

  /**
   * Check if user can manage family members
   */
  static canManageFamily(user: User): boolean {
    return user.role === 'FATHER' || user.role === 'MOTHER';
  }
}
